Add Thrust Slew time to copter

lthall opened the pull request:
This provides a variable that defines the minimum time the thrust can go from zero to full thrust.

This is supposed to be applied after #9381.

This provides a similar feature to that in the solo code except this limits both up going and down going thrust. The variable MOT_SLEW_TIME should be set to approximately 0.275 to be equivalent to that of the solo code. As this implementation is different to the solo code and the solo code also has the motor reset functions we do not guarantee this will address the hardware faults in the same way as the legacy code.

Pedals2Paddles commented:
Awesome, thank you Leonard. I'm not aware of any times when the restart code has been utilized effectively anyway. So I think we're ok without that.

proficnc commented:
Thanks Leonard!

WickedShell commented on libraries/AP_Motors/AP_MotorsCoax.cpp:
```diff
@@ -188,8 +188,14 @@ void AP_MotorsCoax::output_armed_stabilizing()
+    float thrust_yt_cw = thrust_out - 0.5f * yaw_thrust;
+
+    // Apply slew limit to thrust output
+    float thrust_rpyt_out_delta_max = 1.0f/(_slew_time*_loop_rate);
```
divide by zero

WickedShell commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -303,7 +304,15 @@ void AP_MotorsMatrix::output_armed_stabilizing()
+    }
+
+    // apply slew limit to thrust output
+    float thrust_rpyt_out_delta_max = 1.0f/(_slew_time*_loop_rate);
```
divide by zero

WickedShell commented on libraries/AP_Motors/AP_MotorsSingle.cpp:
```diff
@@ -197,7 +197,11 @@ void AP_MotorsSingle::output_armed_stabilizing()
+    float thrust_out = throttle_avg_max + thr_adj;
+
+    // Apply slew limit to thrust output
+    float thrust_rpyt_out_delta_max = 1.0f/(_slew_time*_loop_rate);
```
divide by zero

WickedShell commented on libraries/AP_Motors/AP_MotorsTri.cpp:
```diff
@@ -174,9 +174,15 @@ void AP_MotorsTri::output_armed_stabilizing()
+    float thrust_left = roll_thrust * 0.5f + pitch_thrust * 0.5f;
+    float thrust_rear = pitch_thrust * -0.5f;
+
+    float thrust_rpyt_out_delta_max = 1.0f/(_slew_time*_loop_rate);
```
divide by zero

lthall replied:
Oh Crap, I meant to do that!!!!

OXINARF commented on libraries/AP_Motors/AP_MotorsMulticopter.h:
```diff
@@ -151,6 +152,7 @@ class AP_MotorsMulticopter : public AP_Motors {
     // parameters
     AP_Int16            _yaw_headroom;          // yaw control is given at least this pwm range
     AP_Float            _thrust_curve_expo;     // curve used to linearize pwm to thrust conversion.  set to 0 for linear and 1 for second order approximation
+    AP_Float            _slew_time;     // curve used to linearize pwm to thrust conversion.  set to 0 for linear and 1 for second order approximation
```
Comment is copied.

lthall replied:
Thanks, my bad!

OXINARF commented on libraries/AP_Motors/AP_MotorsCoax.cpp:
```diff
@@ -188,8 +188,19 @@ void AP_MotorsCoax::output_armed_stabilizing()
+    if (!is_zero(_slew_time)) {
+        float thrust_rpyt_out_delta_max = 1.0f/(_slew_time*_loop_rate);
+        _thrust_yt_ccw += constrain_float(_thrust_yt_ccw-thrust_yt_ccw, -thrust_rpyt_out_delta_max, thrust_rpyt_out_delta_max);
+        _thrust_yt_cw += constrain_float(_thrust_yt_cw-thrust_yt_cw, -thrust_rpyt_out_delta_max, thrust_rpyt_out_delta_max);
```
These should be `thrust_yt_ccw - _thrust_yt_ccw` and `thrust_yt_cw - _thrust_yt_cw`. If we dismiss the constrain, it's easily explained. What it is now is:
```
_thrust_yt_ccw = _thrust_yt_ccw + _thrust_yt_ccw - thrust_yt_ccw <=>
_thrust_yt_ccw = 2 * _thrust_yt_ccw - thrust_yt_ccw
```

What we want is:
```
_thrust_yt_ccw = _thrust_yt_ccw + thrust_yt_ccw - _thrust_yt_ccw <=>
_thrust_yt_ccw = thrust_yt_ccw
```

This applies to all changed motor backends.

lthall replied:
Yeh, I am an idiot.

lthall commented:
Ok, I think I have fixed up the lazy formatting and silly mistakes. I also removed the f from the floats as Tridge said we are going to remove that from the code.

OXINARF commented:
> I also removed the f from the floats as Tridge said we are going to remove that from the code.

I'm quite surprised to read that, I would think it was the other way around; this will make the literals be doubles in all platforms except PX4 and ChibiOS (in which, we'll soon want they'll be doubles too).

lthall commented:
Yeh, I tried to go nuts on the formatting so I figured I would do the f thing as well and Randy would leave that out or not.

tridge commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+    thrust_loss_counter++;
+
+    // check if thrust loss for 1 second
+    if (thrust_loss_counter >= (THRUST_LOSS_CHECK_TRIGGER_SEC * scheduler.get_loop_rate_hz())) {
```
this looks like it will flood the log and mavlink

lthall replied:
// exit immediately if thrust boost is already engaged
    if (motors->get_thrust_boost()) {
        return;
    }

That gets prevented by this.

tridge commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+    // check for desired angle over 15 degrees
+    // todo: add thrust angle to AC_AttitudeControl
+    const Vector3f angle_target = attitude_control->get_att_target_euler_cd();
+    if (safe_sqrt(sq(angle_target.x) + sq(angle_target.y)) > THRUST_LOSS_CHECK_ANGLE_DEVIATION_CD) {
```
I don't understand why we're zeroing the counter for this condition

tridge replied:
is the condition backwards?

lthall replied:
This ensures we are descending despite being at maximum throttle and at low thrust angles. We expect to descend when at high bank or pitch angles.

tridge commented on libraries/AP_Motors/AP_MotorsCoax.cpp:
```diff
@@ -188,11 +188,22 @@ void AP_MotorsCoax::output_armed_stabilizing()
+
+    // Apply slew limit to thrust output
+    if (!is_zero(_slew_time)) {
+        float thrust_rpyt_out_delta_max = 1.0 / (_slew_time * _loop_rate);
```
can we use a multiply here?

lthall replied:
I don't know how we can use a multiply there.

OXINARF commented:
@peterbarker Your review applies to PR this builds upon (#9381), as does @tridge's first review. It's better to discuss it there.

lthall commented on ArduCopter/crash_check.cpp:
```diff
@@ -5,6 +5,11 @@
 #define CRASH_CHECK_ACCEL_MAX           3.0f    // vehicle must be accelerating less than 3m/s/s to be considered crashed
 
+// Code to detect a thrust loss main ArduCopter code
+#define THRUST_LOSS_CHECK_TRIGGER_SEC         1     // 1 second decent while level and high throttle indictes thrust loss
```
I am dyslexic so this when right over my head.


but I fixed it with the help of google.

lthall commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+// check for loss of thrust and trigger thrust boost in motors library
+void Copter::thrust_loss_check()
+{
+    static uint16_t thrust_loss_counter;  // number of iterations vehicle may have been crashed
```
thanks

lthall commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+    // check for desired angle over 15 degrees
+    // todo: add thrust angle to AC_AttitudeControl
+    const Vector3f angle_target = attitude_control->get_att_target_euler_cd();
+    if (safe_sqrt(sq(angle_target.x) + sq(angle_target.y)) > THRUST_LOSS_CHECK_ANGLE_DEVIATION_CD) {
```
done

lthall commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+        return;
+    }
+
+    // check for decent
```
Yep, my bad.

lthall commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+    thrust_loss_counter++;
+
+    // check if thrust loss for 1 second
+    if (thrust_loss_counter >= (THRUST_LOSS_CHECK_TRIGGER_SEC * scheduler.get_loop_rate_hz())) {
```
Yep, I don't know how to address this.

lthall commented on ArduCopter/crash_check.cpp:
```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
+    }
+
+    // check for decent
+    if (inertial_nav.get_velocity_z() >= 0.0f) {
```
I have used !is_negative(inertial_nav.get_velocity_z())

Pedals2Paddles commented:
I can say with certainty it explodes with GCS messages warning of thrust loss.  There is definitely a bug in there related to that.  I flipped it over and got hundreds to messages and errors.

peterbarker commented:
On Sat, 6 Oct 2018, Leonard Hall wrote:

> +    // we may have lost thrust
> +    thrust_loss_counter++;
> +
> +    // check if thrust loss for 1 second
> +    if (thrust_loss_counter >= (THRUST_LOSS_CHECK_TRIGGER_SEC * scheduler.get_loop_rate_hz())) {
> 
> Yep, I don't know how to address this.

Since we're just trying to work out how long we've lost thrust for - 
scribble down the time we started to lose thrust (from AP_HAL::millis()) 
and do a subtraction from the current time.  Slight increase in storage 
requirements, probably no slower in terms of execution.

Counters are useful for adding hysteresis - but IIRC we only ever zero the 
counter when the condition is cleared, never decrement it.

lthall commented:
@OXINARF sorry I took some time to work out what you were saying about this pr being based on another. Sorry for the messy way I did this.

@peterbarker , sorry. I do know how to do that I had my head somewhere else. The reason I did it this way is because that is the way the rest of the file is done. I will look at fixing all the instances of this in the file.

@Pedals2Paddles , That is a case I did not consider. We should discuss this over here: (#9381)

tridge commented:
i think the next step is some flight testing, both to look at what the impact is on performance for different values, and to give us confidence it fixes the Solo

Pedals2Paddles commented:
I flew it. It most definitely has an effect. But I'm going to wait for this to be rebased after the thrust loss is merged before trying more.

Pedals2Paddles commented:
Since https://github.com/ArduPilot/ardupilot/pull/9381 has been completed and merged, can this be rebased. I think that would make this PR only be for the slew rate.  I and several others will test once this is rebased.

lthall commented:
Rebase in 20181016 motor slew #9569
